src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs

```diff
@@ -0,0 +1,90 @@
+    {
+        // Symbol is the same for both C and C++, despite the "cpp" or "cc" in file name.
+        // Clang V7 - V9: "safestack.cc.o"
+        // Clang V10 - V14: "safestack.cpp.o"
+        private static readonly string[] symbolSafeStack = new string[] { "safestack.cpp.o", "safestack.cc.o" };
+
+        /// <summary>
+        /// BA3031
+        /// </summary>
+        public override string Id => RuleIds.EnableSafeStackWithClang;
+
+        /// <summary>
+        /// The SafeStack instrumentation pass protects programs by implementing two separate program stacks, 
+        /// one for return addresses and local variables, and the other for everything else.
+        /// To enable SafeStack, pass '-fsanitize=safe-stack' flag to both compile and link command lines.
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.BA3031_EnableSafeStackWithClang_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA3031_Pass),
+                    nameof(RuleResources.BA3031_Error),
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        public override AnalysisApplicability CanAnalyzeElf(ElfBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            IELF elf = target.ELF;
+
+            if (elf.Type == FileType.Core || elf.Type == FileType.None || elf.Type == FileType.Relocatable)
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            if (!target.Compilers.Any(c => c.Compiler == ElfCompilerType.Clang && c.Version.Major >= 7))
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfNotBuiltWithClangV7OrLater;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        /// <summary>
+        /// Checks if SafeStack is enabled by Symbols.
+        /// </summary>
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IELF elf = context.ElfBinary().ELF;
+
+            IEnumerable<ISymbolEntry> symbols =
+                ElfUtility.GetAllSymbols(elf).Where(sym => sym.Type == SymbolType.File);
+
+            if (symbols.Any(s => symbolSafeStack.Contains(s.Name)))
+            {
+                context.Logger.Log(this,
+                       RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                           nameof(RuleResources.BA3031_Pass),
+                           context.TargetUri.GetFileName()));
+            }
+            else
```

Comment:
>else [](http://example.com/codeflow?start=12&length=4)

When you exit the function, you could do a version check, and if it's v3.8 or whatever minimal version we know about, just fire an error.

Otherwise fire a special error that says you should enable SafeStack, you might need to update your version of Clang to do this.

Reply:
agree with this approach, fixed.